design/plugin-backup-and-restore-progress-design.md

```diff
@@ -0,0 +1,238 @@
+	// to user to identify plugin properties related to this backup
+	// +optional
+	PluginSpecific map[string]string `json:"pluginSpecific,omitempty"`
+}
+
+// VolumePluginBackupStatus is the current status of a VolumePluginBackup CR.
+type VolumePluginBackupStatus struct {
+	// Phase is the current state of the VolumePluginBackup.
+	Phase string `json:"phase"`
+
+	// SnapshotID is the identifier for the snapshot of the volume.
+	// This will be used to relate with output in 'velero describe backup'
+	SnapshotID string `json:"snapshotID"`
+
+	// Message is a message about the volume plugin's backup's status.
+	// +optional
+	Message string `json:"message,omitempty"`
+
+	// StartTimestamp records the time a backup was started.
+	// Separate from CreationTimestamp, since that value changes
+	// on restores.
+	// The server's time is used for StartTimestamps
+	// +optional
+	// +nullable
+	StartTimestamp *metav1.Time `json:"startTimestamp,omitempty"`
+
+	// CompletionTimestamp records the time a backup was completed.
+	// Completion time is recorded even on failed backups.
+	// Completion time is recorded before uploading the backup object.
+	// The server's time is used for CompletionTimestamps
+	// +optional
+	// +nullable
+	CompletionTimestamp *metav1.Time `json:"completionTimestamp,omitempty"`
+
+	// Progress holds the total number of bytes of the volume and the current
+	// number of backed up bytes. This can be used to display progress information
+	// about the backup operation.
+	// +optional
+	Progress VolumeOperationProgress `json:"progress,omitempty"`
+}
+
+type VolumeOperationProgress struct {
+	TotalBytes int64
+	BytesDone int64
+}
+
+type VolumePluginBackup struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// +optional
+	Spec VolumePluginBackupSpec `json:"spec,omitempty"`
+
+	// +optional
+	Status VolumePluginBackupStatus `json:"status,omitempty"`
+}
+```
+
+For every backup operation of volume, volume snapshotter creates VolumePluginBackup CR in Velero namespace.
+It keep updating the progress of operation along with other details like Volume name, Backup Name, SnapshotID etc as mentioned in the CR.
+
+In order to know the CR created for the particular backup of a volume, volume snapshotters adds following labels to CR:
+`velero.io/backup-name` with value as Backup Name, and,
+`velero.io/volume-name` with value as volume that is undergoing backup
+Backup name being unique won't cause issues like duplicates in identifying the CR.
+
+Though no restrictions are required on the name of CR, as a general practice, volume snapshotter can name this CR with the value same as return value of CreateSnapshot.
+
+After return from `CreateSnapshot` in `takePVSnapshot`, if VolumePluginBackup CR exists for particular backup of the volume, velero adds this CR to `backupRequest`.
+During persistBackup call, this CR also will be backed up to backup location.
```

Comment:
So these would need to be written in a `json.gz` file, similar to how the snapshots are now?

It may be helpful to create a section in the design for "core" Velero server/client changes. I think the only things that need to be changed are as follows, based on what I've read in this document.

* Creation of the VolumePluginBackup/VolumePluginRestore CRDs at installation time
* Persistence of VolumePluginBackup CRs towards the end of the back up operation
* Sync of VolumePluginBackup/VolumePluginRestore CRDs at the same time as backup synchronization.

Otherwise, volumesnapshotter plugins themselves are responsible for interacting with the VolumePlugin(Backup|Restore) CRs, and Velero's core won't change them.

Then, to see this reported, the `velero describe` commands would need to be aware of the VolumePlugin(Backup|Restore) CRs associated with a Backup CR in order to display the progress fields.

Reply:
sure. I will add a section @nrb. And, the only changes that are required in "core" are those that you mentioned.
Regarding 'velero describe' support for these CRs, if I'm not wrong, 'velero describe' is not aware of PodVolume(Backup|Restore) CRs as well. To keep this PR simple, I thought - we can take it for these new CRs also when PodVolume(Backup|Restore) CRs support is added to 'velero describe'.

Reply:
Currently, no, it would not be aware of these new CRDs, but we could make it support them.

Within the backup describer, we have [a function for describing PodVolumeBackups](https://github.com/vmware-tanzu/velero/blob/master/pkg/cmd/util/output/backup_describer.go#L381), as well as [a function for describing VolumeSnapshotContents](https://github.com/vmware-tanzu/velero/blob/master/pkg/cmd/util/output/backup_describer.go#L493). 

I would think describing these new CRs would work roughly the same way, where they would [get queried from the Kube API](https://github.com/vmware-tanzu/velero/blob/master/pkg/cmd/cli/backup/describe.go#L95) and then [passed to the top level Describe function](https://github.com/vmware-tanzu/velero/blob/master/pkg/cmd/cli/backup/describe.go#L101).

I don't know that all this detail needs to go into the design document, but this is kind of what I envisioned the scope of work to be.

Reply:
As mentioned on Slack, I think the progress tracking is information that's really only necessary when scoped to an in-progress back up operation. When a `Backup` is in a `Completed` or `Failed` state, we know it's at 100% completion.

Given that, does it make sense to synchronize these CRs across clusters?

Reply:
Synchronizing VolumePluginBackup CR to another cluster will help user to know the backup size, number of volumes in it even before user triggers restore of the backup.